src/System.Private.CoreLib/shared/System/ComponentModel/DefaultValueAttribute.cs

```diff
@@ -48,6 +54,40 @@ public DefaultValueAttribute(Type type, string value)
                 {
                     _value = Convert.ChangeType(value, type, CultureInfo.InvariantCulture);
                 }
+
+                return;
+
+                // Try to load and use TypeConverter/TypeDescriptor types
+                bool TryConvertFromInvariantString(Type typeToConvert, string stringValue, out object conversionResult)
+                {
+                    conversionResult = null;
+
+                    // lazy init reflection objects
+                    if (s_getConverterMethodCached == null)
+                    {
+                        Type typeDescriptorType = Type.GetType("System.ComponentModel.TypeDescriptor, System, Version=0.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089", throwOnError: false);
+                        Volatile.Write(ref s_getConverterMethodCached, () => typeDescriptorType?.GetMethod("GetConverter", new Type[] { typeof(Type) }));
```

Comment:
Can we just cache the `MethodInfo`s? Having to re-run `typeDescriptorType?.GetMethod("GetConverter", new Type[] { typeof(Type) }` every time the cache is accessed is not a great cache.

I would just have a `static object s_getConverterMethodCached;` field that is either `null` (which means we didn't populate it), or set to a `MethodInfo` (which means we populated the cache and succeeded), or something else (`new object()`?) (which means we tried to populate it, but couldn't find the method or type - and we should not try again). It's more lightweight than a lambda capture and a delegate.

Most importantly, it makes it easier for static analysis tools to analyze the reflection usage. IL Linkers will typically remove methods that don't appear used, and reflection usage is notoriously hard to statically analyze. Using patterns that are as straightforward as possible makes it less likely these things will get removed. The existing code is passing the type reflected on through a lambda capture and that's one of the harder things to analyze.

Reply:
i followed [this](https://github.com/dotnet/corefx/blob/b9d82ae7fd6912b6c0c758e47d0824689adaeafc/src/System.Runtime.Extensions/src/System/AppDomain.cs#L412) pattern, but there makes more sense Principal could change. Ok thank's for advice!

Reply:
I do not think you should be caching MethodInfos here at all. It should be all just delegates (much faster in steady state than MethodInfo).

Reply:
> It should be all just delegates

could you elaborate? Do you mean use Delegate.Create?

Reply:
Right, like in the Principal example.

Reply:
```cs
...
var converter = getConverter(typeToConvert);
Func<string, object> del = (Func<string, object>)Delegate.CreateDelegate(typeof(Func<string, object>), converter, "ConvertFromInvariantString");
...
```

@jkotas to call instance type TypeConverter with delegate i need to pass concrete converter. What do you think?It's better than invoke(i don't have numbers)?Or we need to cache(key/value) ConvertFromInvariantString delegate?

Reply:
@jkotas @ericstj  i worked on open delegate(first time...very fun, if i could not eat to live i'd work 24/24 on this codebase for free, i'm learning a lot,  exciting...thanks).
I cannot create `Func<object,string,object>` delegate because open delegate creation fail(this could not be object), so i used a on-the-fly generated generics. But now i cannot do a typed invocation but only dynamic one(cannot cast to `Func<object,string,object>`).
What do you think?
```cs
...
Volatile.Write(ref s_convertFromInvariantStringMethod, typeConverterType == null ? new object() : Delegate.CreateDelegate(typeof(Func<,,>).MakeGenericType(typeConverterType, typeof(string), typeof(object)), null, typeConverterType.GetMethod("ConvertFromInvariantString", new Type[] { typeof(string) })));
...
if (!(s_getConverterMethod is Func<Type, object> getConverter) || !(s_convertFromInvariantStringMethod is Delegate convertFromInvariantString))
      return false;

conversionResult = convertFromInvariantString.DynamicInvoke(getConverter(typeToConvert), stringValue);

return true;
```

Reply:
I see. I think the root cause of them problem is that you cannot reference TypeConverter type in CoreLib.

I guess the options to either give up and just use MethodInfo; or add a helper method to System.ComponentModel.TypeConverter that does the whole conversion for you. It is https://github.com/dotnet/corefx/blob/b9d82ae7fd6912b6c0c758e47d0824689adaeafc/src/System.Runtime.Extensions/src/System/AppDomain.cs#L412 has done too - the GetDefaultInstance method is not part of the public surface, it is specific method for use by the AppDomain reflection (https://github.com/dotnet/corefx/blob/master/src/System.Security.Claims/src/System/Security/Claims/GenericPrincipal.cs#L88).

Reply:
> or add a helper method to System.ComponentModel.TypeConverter that does the whole conversion for you

I'll try this way!